usim/_basics/resource.py

```diff
@@ -0,0 +1,208 @@
+
+    def borrow(self, **amounts: T) -> 'BorrowedResources[T]':
+        borrowing = super().borrow(**amounts)
+        assert self._capacity >= borrowing._capacity,\
+            'cannot borrow beyond capacity'
+        return borrowing
+
+
+class Capacities(BorrowedResources[T]):
+    r"""
+    Fixed supply of named resources which can be temporarily borrowed
+
+    The resources and their maximum capacity are defined
+    when the resource supply is created.
+    Afterwards, it is only possible to temporarily :py:meth:`borrow`
+    resources:
+
+    .. code:: python3
+
+        # create a limited supply of resources
+        resources = Capacities(cores=8, memory=16000)
+
+        # temporarily remove resources
+        async with resources.borrow(cores=2, money=4000):
+            await computation
+
+    A :py:class:`~.Capacities` guarantees that its resources are conserved and
+    cannot be leaked. Once resources are :py:meth:`~.borrow`\ ed, they can
+    always be returned promptly.
+    """
+    def __init__(self, __zero__: Optional[T] = None, **capacity: T):
+        resources = Resources(__zero__, **capacity)
+        super().__init__(resources, resources.levels)
+        self._available = Tracked(resources.levels)
+
+
+class Resources(BaseResources[T]):
+    r"""
+    Supply of named resources which can be temporarily borrowed or produced/consumed
+
+    The resources and their initial levels are defined
+    when the resource supply is created.
+    Afterwards, the level of resources can be permanently :py:meth:`~.increase`\ d or
+    :py:meth:`~.decrease`\ d as well as temporarily decreased by :py:meth:`borrow`\ ing:
+
+    .. code:: python3
+
+        # create an open supply of resources
+        resources = Resources(cores=8, memory=4000)
+
+        # increase the resource supply available
+        resources.increase(memory=2000)
+
+        # temporarily remove resources
+        async with resources.borrow(cores=2, memory=6000):
+            await computation
+
+        # decrease the resource supply available
+        resources.decrease(cores=4)
+
+    A :py:class:`~.Capacities` guarantees that is always possible
+    to increase the level of available resources.
+    Once resources are :py:meth:`~.borrow`\ ed, they can
+    always be returned promptly.
+    """
+    def __init__(self, __zero__: Optional[T] = None, **capacity: T):
+        if not capacity:  # Note: this should be a type-error not assert for consistency
+            raise TypeError(
+                '%s requires at least 1 keyword-only argument' % self.__class__.__name__
+            )
+        __zero__ = __zero__ if __zero__ is not None else\
+            type(next(iter(capacity.values())))()  # bare type invocation must be zero
+        self._levels_type = __specialise__(__zero__, capacity.keys())
+        self._available = Tracked(self._levels_type(**capacity))
+        assert self._available > self._levels_type.zero,\
```

Comment:
Shouldn't this also allow for `0` values? If I imagine a producing resource it might start with `0`. Let us imaging a pizza bakery, if there are no orders, no pizza will be available.

Reply:
Yes, good point. Same for other actions, I assume.

Reply:
```suggestion
        assert self._available >= self._levels_type.zero,\
```